Bugfix for writer not being able to open again after closing

yschulz opened the pull request:
- Fixes #1598
- After discussion in #1590 I verified the issue in and reapplied the changes and comments to `rolling`. To reiterate, the `storage_factory` needs to persist beyond the call to close a current bag. On the other hand, all containers that were built during the recording have to be cleared before opening another bag.

yschulz committed:
re-applies fixes from #1590 to rolling. Also removes new message definition in sequential writer test for multiple open operations. Also clears topic_names_to_message_definitions_ and handles message_definitions_s underlying container similarly. Lastly, also avoids reset of factory in the compression writer, adds unit test there too.

Signed-off-by: Yannick Schulz <yschulz854@gmail.com>

yschulz commented:
I keep it as a draft until @MichaelOrlov and @fujitatomoya can verify what all can be attached to a writers lifetime and what needs to be reset. I am especially not sure about the compressed writer, since I haven't spend enough time looking at that one. Thank you guys for your comments and review, I guess we can close the other PR?

fujitatomoya commented:
@yschulz so this replaces https://github.com/ros2/rosbag2/pull/1590, right? in that case, you can close https://github.com/ros2/rosbag2/pull/1590, and let us know this one is ready to review.

yschulz committed:
removes unused compressor_ member from compresser writer class. Also delegates rest of the closing behavior to the base class in close method, as it is handled in the open and write methods of the compression writer

Signed-off-by: Yannick Schulz <yschulz854@gmail.com>

yschulz commented:
@fujitatomoya done, and yes please :)

I had a closer look at the compression writer and as to my impression, there is no need to clear anything regarding the compression functionality. However, I saw that the closing of all the storage functionality can be directly delegated to the base class (as it is for opening and writing anyway). Also, spotted an unused member, maybe a relict from before the threaded approach, you may know more.

MichaelOrlov commented on rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp:
```diff
@@ -164,7 +164,7 @@ class ROSBAG2_CPP_PUBLIC SequentialWriter
   std::mutex topics_info_mutex_;
 
-  LocalMessageDefinitionSource message_definitions_;
+  std::unique_ptr<LocalMessageDefinitionSource> message_definitions_;
```
There are no benefits to making `message_definitions_` a unique pointer and resetting it on each writer's open/close operations. I would say doing so will slow down consequent write operations to the newly open writer in comparison to the original implementation.
Please return it back to reduce unrelated delta.

MichaelOrlov replied:
The message_definitions_  was addressed in the commit https://github.com/ros2/rosbag2/pull/1599/commits/5a1df9e50c6b10964b4b73951613b5f7b4a8c144

MichaelOrlov commented on rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp:
```diff
@@ -69,7 +70,9 @@ SequentialWriter::~SequentialWriter()
   // destructed objects.
   callback_manager_.delete_all_callbacks();
-  close();
+  if (storage_) {
+    close();
```
Relying on the assumption that `storage_` object indicates that the writer is open or has been closed is incorrect here.
The reason for that is for example we can fail to create a new storage in the `bag_split(..)` method and throw an exception. However, we want ti try to write updated `metadata.yaml` file on the destruction of the `writer` object.

The better way would be to create a separate variable protected by a mutex to indicate if the writer is open or closed.

MichaelOrlov replied:
Fixed in the commit: https://github.com/ros2/rosbag2/pull/1599/commits/0bd5003e2cb1ca2235580009ef10b2729838c608

MichaelOrlov commented on rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp:
```diff
@@ -186,7 +192,12 @@ void SequentialWriter::close()
+  topics_names_to_info_.clear();
+  topic_names_to_message_definitions_.clear();
+
+  converter_.reset();
+  message_definitions_.reset();
```
```suggestion
```

The   `converter_` and `message_definitions_.` is unrelated for the fix and will be beneficial to let them persist over open/close operations.

MichaelOrlov replied:
We actually need to reset the converter on close. I was wrong.
The message_definitions_  was addressed in the commit https://github.com/ros2/rosbag2/pull/1599/commits/5a1df9e50c6b10964b4b73951613b5f7b4a8c144

morlov-apexai committed:
Remove unrelated delta

- message_definitions_ was intentionally allocated on the stack and
should persist between writer close() and open() because it represents
cache for message definitions which is not changes.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Don't call virtual methods from destructors

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Cleanup 'rosbag2_directory_next' after the test run

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Protect Writer::open(..) and Writer::close() with mutex on upper level

- Rationale: To avoid race conditions if open(..) and close() could be
ever be called from different threads.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Bugfix for WRITE_SPLIT callback not called for the last compressed file

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Bugfix for lost messages from cache when closing compression writer

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Use dedicated is_open_ flag instead of relying on storage_

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

yschulz commented:
Thanks for the review and addressing the changes directly.

MichaelOrlov commented:
@yschulz Sorry for taking a long time to review. It was a rush time before Jazzy code freeze. And changes per this PR were non-trivial for review. Was need to verify that we will not break anything.

MichaelOrlov commented:
Gist: https://gist.githubusercontent.com/MichaelOrlov/2677c3394586488cd9a777215205eff9/raw/4e54566b952d8218af61493f02a1e01394e567bf/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_cpp rosbag2_compression rosbag2_tests
TEST args:  --packages-above rosbag2_cpp rosbag2_compression rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13815
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=20992)](http://ci.ros2.org/job/ci_linux/20992/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15294)](http://ci.ros2.org/job/ci_linux-aarch64/15294/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21673)](http://ci.ros2.org/job/ci_windows/21673/)

fujitatomoya commented:
@MichaelOrlov thanks for checking this, i will take a look at this today.

MichaelOrlov commented:
Re-run Windows CI
It seems the previous failure is unrelated to the changes from this PR.
```
C:\ci\ws\src\yschulz\rosbag2\rosbag2_storage\include\rosbag2_storage/yaml.hpp(28,10): fatal error C1083: Cannot open include file: 'yaml-cpp/yaml.h': No such file or directory (compiling source file C:\ci\ws\src\yschulz\rosbag2\rosbag2_storage\src\rosbag2_storage\metadata_io.cpp) [C:\ci\ws\build\rosbag2_storage\rosbag2_storage.vcxproj]
```
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21677)](http://ci.ros2.org/job/ci_windows/21677/)

MichaelOrlov commented:
@clalancette Out of curiosity if this is a known issue 
```
C:\ci\ws\src\yschulz\rosbag2\rosbag2_storage\include\rosbag2_storage/yaml.hpp(28,10): fatal error C1083: Cannot open include file: 'yaml-cpp/yaml.h': No such file or directory (compiling source file C:\ci\ws\src\yschulz\rosbag2\rosbag2_storage\src\rosbag2_storage\metadata_io.cpp) [C:\ci\ws\build\rosbag2_storage\rosbag2_storage.vcxproj]
```
Or we missing something like `find_package(yaml-cpp REQUIRED)` in CMake file?

clalancette commented:
> @clalancette Out of curiosity if this is a known issue

No, not a known issue.  This succeeded on the nightlies: https://ci.ros2.org/view/nightly/job/nightly_win_rel/3037/ .

My best guess is that this PR needs to be rebased onto the latest rosbag2 code to get #1605

fujitatomoya reviewed (approved):
lgtm with green CI

MichaelOrlov commented:
Re-run Windows CI after rebasing on the latest Roling
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21679)](http://ci.ros2.org/job/ci_windows/21679/)

MichaelOrlov commented:
Ok. CI is green after rebase. Merging then.

MichaelOrlov commented:
https://github.com/Mergifyio backport jazzy
